src/Microsoft.ML.Transforms/Text/WordEmbeddingsTransform.cs

```diff
@@ -0,0 +1,448 @@
+            }
+
+            Host.CheckNonWhiteSpace(_modelFileNameWithPath, nameof(_modelFileNameWithPath));
+            _currentVocab = GetVocabularyDictionary();
+            _outputType = new VectorType(NumberType.R4, 3 * _currentVocab.Dimension);
+            Metadata.Seal();
+        }
+
+        public static WordEmbeddingsTransform Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            IHost h = env.Register(RegistrationName);
+            h.CheckValue(ctx, nameof(ctx));
+            h.CheckValue(input, nameof(input));
+            return h.Apply("Loading Model",
+                ch => new WordEmbeddingsTransform(h, ctx, input));
+        }
+
+        public override void Save(ModelSaveContext ctx)
+        {
+            Host.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            SaveBase(ctx);
+            ctx.Writer.WriteBoolByte(_customLookup);
+            if (_customLookup)
+                ctx.SaveString(_modelFileNameWithPath);
+            else
+                ctx.Writer.Write((uint)_modelKind);
+        }
+
+        protected override ColumnType GetColumnTypeCore(int iinfo)
+        {
+            Host.Assert(0 <= iinfo && iinfo < Infos.Length);
+            return _outputType;
+        }
+
+        protected override Delegate GetGetterCore(IChannel ch, IRow input, int iinfo, out Action disposer)
+        {
+            Host.AssertValue(ch);
+            ch.AssertValue(input);
+            ch.Assert(0 <= iinfo && iinfo < Infos.Length);
+            disposer = null;
+
+            var info = Infos[iinfo];
+            if (!info.TypeSrc.IsVector)
+            {
+                throw Host.ExceptParam(nameof(input),
+                    "Text input given, expects a text vector");
+            }
+            return GetGetterVec(ch, input, iinfo);
+        }
+
+        private ValueGetter<VBuffer<float>> GetGetterVec(IChannel ch, IRow input, int iinfo)
+        {
+            Host.AssertValue(ch);
+            ch.AssertValue(input);
+            ch.Assert(0 <= iinfo && iinfo < Infos.Length);
+
+            var info = Infos[iinfo];
+            ch.Assert(info.TypeSrc.IsVector);
+            ch.Assert(info.TypeSrc.ItemType.IsText);
+
+            var srcGetter = input.GetGetter<VBuffer<DvText>>(info.Source);
+            var src = default(VBuffer<DvText>);
+            int dimension = _currentVocab.Dimension;
+            float[] wordVector = new float[_currentVocab.Dimension];
+
+            return
+                (ref VBuffer<float> dst) =>
+                {
+                    int deno = 0;
+                    srcGetter(ref src);
+                    var values = dst.Values;
+                    Utils.EnsureSize(ref values, 3 * dimension, keepOld: false);
```

Comment:
With `keepOld: false`, the `Utils.EnsureSize()` function always allocates a new array. Would it be faster to perform the size check and just zero the existing?

https://github.com/dotnet/machinelearning/blob/c023727b76970ab913ec1ce38276508835c17bcf/src/Microsoft.ML.Core/Utilities/Utils.cs#L864

Reply:
I'm actually not sure why we allocate new array all the time, (considering what we clean it immediately after) so I rather remove this "keepOld"

---
In reply to: [204564374](https://github.com/dotnet/machinelearning/pull/545#discussion_r204564374) [](ancestors = 204564374)

Reply:
Reactivating. Read implementation of `EnsureSize` a bit more carefully. THe `keepOld: false` certainly does not always allocate a new array. The difference is, in the situation where it *is* necessary to resize the array, whether that is created through `Array.Resize` or `new`... the latter is faster.

---
In reply to: [204578950](https://github.com/dotnet/machinelearning/pull/545#discussion_r204578950) [](ancestors = 204578950,204564374)